modules/nback.llvm/llvm.bal

```diff
@@ -77,31 +85,84 @@ public distinct class BasicBlock {
     }
 }
 
+public type LinkageType "internal"|"external";
+
+# Corresponds to llvm::FunctionType class
+public type FunctionType record {|
+    Type returnType;
+    Type[] paramTypes = [];
+|};
+
+# Corresponds to llvm::Function class
 public class Function {
     private BasicBlock[] basicBlocks = [];
     private int varCount = 0;
     private int labelCount = 0;
+    private string functionName;
+    private Type returnType;
+    private Value[] paramValues;
+    private LinkageType linkageType = "external";
 
     // XXX need stuff for the definition
-    public function init() {
+    function init(string functionName, FunctionType functionType) {
+        self.functionName = functionName;
+        self.returnType = functionType.returnType;
+        self.paramValues = [];
+        foreach var paramType in functionType.paramTypes {
+            string register = self.genReg();
+            Value arg = new (paramType, register);
+            self.paramValues.push(arg);
+        }
+    }
+
+    public function paramByIndex(int index) returns Value {
+        return self.paramValues[index];
+    }
 
+    public function setLinkageType(LinkageType linkageType) {
+        self.linkageType = linkageType;
     }
 
-    public function outputBody(Output out) {
-        string[] lines = [];
+    public function output(Output out) {
+        out.push(self.header());
+        self.outputBody(out);
+        out.push("}");
+    }
+
+    function header() returns string {
+        string[] words = [];
+        words.push("define");
+        if self.linkageType != "external" {
+            words.push(self.linkageType);
+        }
+        words.push(serializeType(self.returnType));
+        words.push(self.functionName);
```

Comment:
The API user should not supply the `@`: they should give the actual name of the function

Reply:
Fixed